File: java/01-bowling/src/test/java/bowling/GameTest.java
package bowling;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class GameTest {

    private Game game = new Game();

    @Test
    public void gutterGame() {
        rollMany(0);
        assertEquals(0, game.calculateScore());
    }

    @Test
    public void allOnes() {
        rollMany(1);
        assertEquals(20, game.calculateScore());
    }

    @Test
    public void spare() {
        rollSpare(game);
        game.roll(3);
        assertEquals(16, game.calculateScore());
    }

    @Test
    public void strike() {
        rollStrike();
        game.roll(4);
        game.roll(5);
        assertEquals(28, game.calculateScore());
    }

    @Test
    public void perfectGame() {
        rollMany(12, 10);
        assertEquals(300, game.calculateScore());
    }

    private void rollStrike() {
        game.roll(10);
    }

    private void rollSpare(Game game) {
        game.roll(5);
        game.roll(5);
    }

    private void rollMany(int numPins) {
        rollMany(20, numPins);
    }

    private void rollMany(int times, int numPins) {
        for (int i = 0; i < times; i++) {
            game.roll(numPins);
        }
    }
}

File: kotlin/01-bowling/src/test/kotlin/bowling/GameTest.kt
package bowling

import org.junit.Assert.assertEquals
import org.junit.Test

class GameTest {
    private val game = Game()

    @Test
    fun gutterGame() {
        game.rollMany(0)
        assertEquals(0, game.score)
    }

    @Test
    fun allOnes() {
        game.rollMany(1)
        assertEquals(20, game.score)
    }

    @Test
    fun spare() {
        game.rollSpare()
        game.roll(3)
        assertEquals(16, game.score)
    }

    @Test
    fun strike() {
        game.rollStrike()
        game.roll(5)
        game.roll(3)
        assertEquals(26, game.score)
    }

    @Test
    fun perfectGame() {
        game.rollMany(10, 12)
        assertEquals(300, game.score)
    }

    private fun Game.rollSpare() {
        roll(5)
        roll(5)
    }

    private fun Game.rollMany(numPins: Int, numRolls: Int = 20) {
        for (i in 1..numRolls) {
            roll(numPins)
        }
    }

    private fun Game.rollStrike() {
        roll(10)
    }
}
